paths_cli/compiling/core.py

```diff
@@ -0,0 +1,300 @@
+    """
+    SCHEMA = "http://openpathsampling.org/schemas/sim-setup/draft01.json"
+    category = None
+    def __init__(self, builder, parameters, name=None, aliases=None,
+                 requires_ops=(1, 0), requires_cli=(0, 3)):
+        super().__init__(requires_ops, requires_cli)
+        self.aliases = aliases
+        self.attribute_table = {p.name: p.loader for p in parameters
+                                if p.required}
+        self.optional_attributes = {p.name: p.loader for p in parameters
+                                    if not p.required}
+        self.defaults = {p.name: p.default for p in parameters
+                         if not p.required}
+        self.name = name
+        self.builder = builder
+        self.builder_name = str(self.builder)
+        self.parameters = parameters
+        logger_name = f"compiler.InstanceCompilerPlugin.{builder}"
+        self.logger = logging.getLogger(logger_name)
+
+    @property
+    def schema_name(self):
+        if not self.name.endswith(self.category):
+            schema_name = f"{self.name}-{self.category}"
+        else:
+            schema_name = self.name
+        return schema_name
+
+    def to_json_schema(self, schema_context=None):
+        parameters = dict(p.to_json_schema() for p in self.parameters)
+        properties = {
+            'name': {'type': 'string'},
+            'type': {'type': 'string',
+                     'enum': [self.name]},
+        }
+        properties.update(parameters)
+        required = [p.name for p in self.parameters if p.required]
+        dct = {
+            'properties': properties,
+            'required': required,
+        }
+        return self.schema_name, dct
+
+    def compile_attrs(self, dct):
+        """Compile the user input dictionary to mapping of name to object.
+
+        This changes the values in the key-value pairs we get from the file
+        into objects that are suitable as input to a function. Further
+        remapping of keys is performed by the builder.
+
+        Parameters
+        ----------
+        dct: Dict
+            User input dictionary (from YAML, JSON, etc.)
+
+        Returns
+        -------
+        Dict :
+            Mapping with the keys from the input dictionary, but values are
+            now appropriate inputs for the builder.
+        """
+        # TODO: support aliases in dct[attr]
+        input_dct = self.defaults.copy()
+        self.logger.debug(f"defaults: {input_dct}")
+        input_dct.update(dct)
+        self.logger.debug(f"effective input: {input_dct}")
+
+        new_dct = {}
+        for attr, func in self.attribute_table.items():
+            try:
+                value = input_dct[attr]
+            except KeyError:
+                raise InputError(f"'{self.builder_name}' missing required "
+                                 f"parameter '{attr}'")
+            self.logger.debug(f"{attr}: {input_dct[attr]}")
+            new_dct[attr] = func(input_dct[attr])
```

Comment:
value should be valid at the end
```suggestion
            try:
                value = input_dct[attr]
            except KeyError:
                raise InputError(f"'{self.builder_name}' missing required "
                                 f"parameter '{attr}'")
            self.logger.debug(f"{attr}: {value}")
            new_dct[attr] = func(value)
```